### PR TITLE
Haddock docs, minor `examples` reorganisation

### PR DESCRIPTION
**Pull request by t1lde**
Added some haddock docs, and some (small) relevant examples.

WIP - some more things could use doc comments, and the `Examples` modules may not be the best place to keep the examples (would be nice to better expose them to haddock...)

### CODE REVIEW

**Comment by L-as** [comment]
Why is this a draft?

File: examples/Examples/PlutusType.hs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  A Sum type, which can be encoded as an Enum
+-}
+data AB (s :: k) = A | B
+  
+{- | 
+  AB is encoded as an Enum, using values of PInteger
+  internally.
+-}
+instance PlutusType AB where
+  type PInner AB _ = PInteger
```
Comment (L-as) [inline examples/Examples/PlutusType.hs:22]: This is quite clever.

**Comment by t1lde** [comment]
Put this up as draft before I logged off last night, but it should be ready ig.

Will fix formatting.

**Comment by t1lde** [comment]
Looks like there's a fourmolu idempotentcy bug, something to do with Haddock that appears to be fixed in fourmolu 0.4.0.0, which hasn't made it's way into nixpkgs yet :upside_down_face:

**Comment by L-as** [comment]
We need to use `fourmolu_0_4_0_0` instead of `fourmolu` in our Flake.

**Comment by L-as** [comment]
See 8e815b11eed5ac324e570a88647ae3af574d5c79 from tmp branch. You can `git reset` to this.

**Comment by L-as** [comment]
Nevermind, CI still fails...

**Comment by L-as** [comment]
See 47775379dacb21944043bc3ea044cb18cf1bca3d instead, apparently `fourmolu_0_4_0_0` is broken on NIxpkgs. We could include it using `haskell.nix`, but :shrug: for now.

File: Plutarch.hs [hunk 2]
```diff
@@ -106,30 +173,64 @@ instance PlutusType a => PCon a where
   pcon = punsafeCoerce . pcon'
 
 class PCon a where
+  -- | Construct a Plutarch Term via a Haskell value
   pcon :: a s -> Term s a
 
 class PMatch a where
+  -- | Construct a Plutarch Term via a Haskell value
```
Comment (TotallyNotChase) [inline Plutarch.hs:180]: I think this one needs to be updated.

**Review by TotallyNotChase — CHANGES_REQUESTED** [review]
Looks great! The haddocks and examples are going to very useful for #55 as well.

File: Plutarch.hs (hunk 2 above)
Comment (t1lde) [reply Plutarch.hs]: Done

**Comment by L-as** [comment]
I just noticed you made this merge into `staging`, but `staging` is just my personal playground currently... this is also on me lol.

**Pull request merged**